ldms/src/sampler/coretemp/coretemp.c

```diff
@@ -0,0 +1,424 @@
+{
+	int rc;
+	regmatch_t match[16];
+	char path[PATH_MAX];
+	char *coretemp = strstr(fpath, "coretemp");
+	if (!coretemp)
+		return 0;
+	switch (typeflag) {
+	case FTW_F:  /* fpath is a regular file. */
+		rc = regexec(&sensor_re, fpath, 16, match, 0);
+		if (!rc) {
+			struct sensor *sensor = calloc(1, sizeof *sensor);
+			sensor->key.package_no = strtod(&fpath[match[2].rm_so], NULL);
+			sensor->key.sensor_no =	strtod(&fpath[match[7].rm_so], NULL);
+			char *dirn = strdup(fpath);
+			dirn = dirname(dirn);
+			sprintf(path, "%s/temp%d_label",  dirn, sensor->key.sensor_no);
+			sensor->temp_label_path = strdup(path);
+			sprintf(path, "%s/temp%d_max", dirn, sensor->key.sensor_no);
+			sensor->temp_max_path = strdup(path);
+			sprintf(path, "%s/temp%d_crit", dirn, sensor->key.sensor_no);
+			sensor->temp_crit_path = strdup(path);
+			sprintf(path, "%s/temp%d_crit_alarm", dirn, sensor->key.sensor_no);
+			sensor->temp_crit_alarm_path = strdup(path);
+			sprintf(path, "%s/temp%d_input", dirn, sensor->key.sensor_no);
+			sensor->temp_input_path = strdup(path);
+			rbn_init(&sensor->rbn, &sensor->key.package_no);
+			rbt_ins(&sensor_tree, &sensor->rbn);
+		}
+		break;
+	case FTW_D:  /* fpath is a directory. */
+		break;
+	case FTW_DNR: /* fpath is a directory which can't be read. */
+	case FTW_NS: /* The stat(2) call failed on fpath, which is not a symbolic link. */
+	case FTW_SL: /* Symbolic link.  */
+	default:
+		break;
+	}
+	return 0;
+}
+
+void collect_sensors(void)
+{
+	struct rbn *rbn;
+	struct sensor *sens;
+	FILE *fp;
+	char fbuf[80];
+	int rc;
+
+	RBT_FOREACH(rbn, &sensor_tree) {
+		sens = container_of(rbn, struct sensor, rbn);
+		/* temp_label -- doens't change */
+		if (!sens->temp_label) {
+			fp = fopen(sens->temp_label_path, "r");
+			rc = fread(fbuf, 1, sizeof(fbuf), fp);
+			fclose(fp);
+			char *nl = strstr(fbuf, "\n");
+			if (nl)
+				*nl = '\0';
+			fbuf[rc] = '\0';
+			sens->temp_label = strdup(fbuf);
+		}
+		/* temp_input */
+		if (!sens->temp_input_fp)
+			sens->temp_input_fp = fopen(sens->temp_input_path, "r");
+		else
+			fseek(sens->temp_input_fp, 0, SEEK_SET);
+		rc = fread(fbuf, 1, sizeof(fbuf), sens->temp_input_fp);
+		fbuf[rc] = '\0';
+		sens->temp_input = (float)strtol(fbuf, NULL, 0) / 1000.0;
+		/* temp_max */
+		if (!sens->temp_max_fp)
+			sens->temp_max_fp = fopen(sens->temp_max_path, "r");
+		else
+			fseek(sens->temp_max_fp, 0, SEEK_SET);
+		rc = fread(fbuf, 1, sizeof(fbuf), sens->temp_max_fp);
+		fbuf[rc] = '\0';
+		sens->temp_max = (float)strtol(fbuf, NULL, 0) / 1000.0;
+		/* temp_crit */
+		if (!sens->temp_crit) {
+			fp = fopen(sens->temp_crit_path, "r");
+			rc = fread(fbuf, 1, sizeof(fbuf), fp);
+			fclose(fp);
+			fbuf[rc] = '\0';
+			sens->temp_crit = (float)strtol(fbuf, NULL, 0) / 1000;
+		}
+		/* temp_crit_alarm */
+		if (!sens->temp_crit_alarm_fp)
+			sens->temp_crit_alarm_fp = fopen(sens->temp_crit_alarm_path, "r");
```

Comment:
Similar to `sens->temp_max_fp`, should we fclose `sens->temp_crit_alarm_fp` after the loop?

Reply:
This is the same as the other one. The reason some files are closed is because their contents never change, e.g. temp_label; which is the text label for the temperature sensor and it never changes. So that file is opened once, read, and then closed. It is never read again. By contrast temp_crit_alarm_fp may change, so it is opened once, and then subsequently fseek(0) before reading.